content/blog/learn-pulumi-inputs-outputs/index.md

```diff
@@ -0,0 +1,160 @@
+    const alb = new awsx.lb.ApplicationLoadBalancer("web-traffic", { securityGroups: [ sg ] });
+    const listener = alb.createListener("web-listener", { port: 80 });
+    const publicIps: pulumi.Output<string>[] = [];
+    const subnets = await vpc.publicSubnets;
+```
+
+It's important to remember you can't access the values of outputs directly. They might not be known, so for example, printing the value to the console returns an error instead. The most common way of interacting with an output is to pass it to another resource as an input.
+
+In this example,  we've instantiated a virtual machine per subnet, and we're passing the public IP address as its target.
+
+```typescript
+    // For each subnet, and each subnet/zone, create a VM and a listener.
+    for (let i = 0; i < subnets.length; i++) {
+        // Create the instance in the same VPC
+        const vm = new aws.ec2.Instance(`web-${i}`, {
+            ami: aws.getAmi({
+                filters: [
+                    { name: "name", values: [ "ubuntu/images/hvm-ssd/ubuntu-trusty-14.04-amd64-server-*" ] },
+                    { name: "virtualization-type", values: [ "hvm" ] },
+                ],
+                mostRecent: true,
+                owners: [ "099720109477" ], // Canonical
+            }).then(ami => ami.id),
+            instanceType: "m5.large",
+            subnetId: subnets[i].subnet.id,
+            availabilityZone: subnets[i].subnet.availabilityZone,
+            vpcSecurityGroupIds: alb.securityGroups.map(sg => sg.securityGroup.id),
+            userData: `#!/bin/bash
+            echo "Hello World, from Server ${i+1}!" > index.html
+            nohup python -m SimpleHTTPServer 80 &`,
+        }, providerOpts);
+        publicIps.push(vm.publicIp);
+        alb.attachTarget("target-" + i, vm);
+    }
+
+    // Export the resulting URL so that it's easy to access.
+    return { endpoint: listener.endpoint.hostname, publicIPs: publicIps  };
+};
+```
+
+As the diagram earlier depicted, another thing we can easily do with outputs is export them. When we export an output, Pulumi captures that value and makes it easily available on the stack. This includes at the command line, in the pulumi console, and even if you want one stack to depend on another stack's output. You can see it in action with the following script:
+
+```bash
+for i in {1..25}; do curl http://$(pulumi stack output endpoint); done
+```
+
+Let's look at other ways of interacting with the concrete value of an output. In this example, we want to parse out the subdomain of the load balancer's hostname. Usually, you would expect a string, and we can call functions like substring and indexOf directly, but in this case, we can't because it's an Output. This is where we can use the apply function. Pulumi invokes a callback when it knows the concrete value and apply takes the callback and returns another output that we can export.
```

Comment:
```suggestion
Let's look at other ways of interacting with the concrete value of an output. In this example, we want to parse out the subdomain of the load balancer's hostname. Usually, you would expect a string, and we can call functions like `substring()` and `indexOf()` directly, but in this case, we can't because it's an Output. This is where we can use the `apply` function. Pulumi invokes a callback when it knows the concrete value, and `apply` takes the callback and returns another output that we can export.
```

We should link to this (to apply) in the docs.